content/topics/clean-code/python/_index.md

```diff
@@ -0,0 +1,174 @@
+---
+title: Clean Code for Python
+ready: True
+---
+
+# Clean Code for Python
+
+## Tabs or Spaces
+
+`4 spaces == 1 tab`
+
+Spaces are the preferred indentation method.
+Tabs should be used solely to remain consistent with code that is already indented with tabs.
+Python 3 **DISALLOWS** mixing the use of tabs and spaces for indentation.
+Python 2 code indented with a mixture of tabs and spaces should be converted to using spaces exclusively.
+
+Yes:
+
+    # Aligned with opening delimiter.
+    foo = long_function_name(var_one, var_two,
+                            var_three, var_four)
+
+    # Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest.
+    def long_function_name(
+            var_one, var_two, var_three,
+            var_four):
+        print(var_one)
+
+    # Hanging indents should add a level.
+    foo = long_function_name(
+        var_one, var_two,
+    var_three, var_four)
+
+No:
+
+    # Arguments on first line forbidden when not using vertical alignment.
+    foo = long_function_name(var_one, var_two,
+        var_three, var_four)
+
+    # Further indentation required as indentation is not distinguishable.
+    def long_function_name(
+        var_one, var_two, var_three,
+        var_four):
+        print(var_one)
+
+## Imports
+
+- Imports should usually be on separate lines:
+
+```
+  Yes: import os
+  import sys
+
+  No: import sys, os
+```
+
+It's okay to say this though:
+
+```
+from subprocess import Popen, PIPE
+```
+
+- Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants. Imports should be grouped in the following order:
+
```

Comment:
This can break test mocks. It is deep magic.

In general I import stuff inside the function that needs it unless it is used all over the place. It acts as a form of documentation and is more robust. And the code can also run quicker as a result